### ISSUE TITLE
`datafusion_proto` crate should have exhaustive match statements for handling `Expr`

### ISSUE THREAD

**Issue by andygrove** [apache/arrow-datafusion#2565]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
When adding new `Expr` variants, we should also be adding code to `datafusion_proto` to handle them but we are not forced into doing this today because the match statement handles the `_` case with an error.

**Describe the solution you'd like**
Support all current `Expr` variants in and remove the `_` case.

**Describe alternatives you've considered**
None

**Additional context**
None

**Comment by andygrove** [comment]
Currently `datafusion.proto` only handles expressions and not logical plans, but that will have to change if we want to support subquery expressions. Ballista already has support for plans so maybe we move some of that code to DataFusion? Another option would be to look at Substrait but that may be a significant amount of work.

fyi @alamb @carols10cents @Ted-Jiang @thinkharderdev since you have contributed to this file

**Comment by alamb** [comment]
I would think we should move the code for serializing logical plans into datafusion.proto as well

**Comment by thinkharderdev** [comment]
Yeah, agreed. I think it makes sense to move all serialization into datafusion-proto. It might be useful as well if we move forward with breaking Ballista out into it's own repo to help ensure that constructs added to DataFusion are always serializable.

**Comment by Ted-Jiang** [comment]
+1 , remove the `_` case, will give error at compile time.
